Backported from 1.9.4 to 1.9.3:

From 43ca54a869ddea7085c9526ac2c69e26f446f817 Mon Sep 17 00:00:00 2001
From: Daniel Stone <daniel@fooishbar.org>
Date: Thu, 10 Feb 2011 17:32:34 +0000
Subject: Input: Store passive-activating key in CheckDeviceGrabs

CheckDeviceGrabs will activate a passive grab for KeyPress and
ButtonPress events.  GrabInfoRec::activatingKey contains the keycode
which activated the passive grab, so we can deactivate it later in
ProcessOtherEvents.

Previously, CheckDeviceGrabs relied on its callers to set
activatingKey, which not all callers were doing (I'm looking at you,
ComputeFreezes).  Just set it in CheckDeviceGrabs instead.

Signed-off-by: Daniel Stone <daniel@fooishbar.org>
---
diff -upr xorg-server-1.9.3.orig/Xi/exevents.c xorg-server-1.9.3.work/Xi/exevents.c
--- xorg-server-1.9.3.orig/Xi/exevents.c	2010-12-22 18:58:55.033527000 -0800
+++ xorg-server-1.9.3.work/Xi/exevents.c	2011-02-18 18:43:48.753214000 -0800
@@ -1053,10 +1053,8 @@ ProcessOtherEvent(InternalEvent *ev, Dev
     switch(event->type)
     {
         case ET_KeyPress:
-            if (!grab && CheckDeviceGrabs(device, event, 0)) {
-                device->deviceGrab.activatingKey = key;
+            if (!grab && CheckDeviceGrabs(device, event, 0))
                 return;
-            }
             break;
         case ET_KeyRelease:
             if (grab && device->deviceGrab.fromPassiveGrab &&
diff -upr xorg-server-1.9.3.orig/dix/events.c xorg-server-1.9.3.work/dix/events.c
--- xorg-server-1.9.3.orig/dix/events.c	2010-12-22 18:58:55.512528000 -0800
+++ xorg-server-1.9.3.work/dix/events.c	2011-02-18 18:48:08.259228000 -0800
@@ -3620,6 +3620,7 @@ CheckDeviceGrabs(DeviceIntPtr device, De
     WindowPtr pWin = NULL;
     FocusClassPtr focus = IsPointerEvent((InternalEvent*)event) ? NULL : device->focus;
     BOOL sendCore = (IsMaster(device) && device->coreEvents);
+    Bool ret = FALSE;
 
     if (event->type != ET_ButtonPress &&
         event->type != ET_KeyPress)
@@ -3638,14 +3639,17 @@ CheckDeviceGrabs(DeviceIntPtr device, De
 	    pWin = focus->trace[i];
 	    if (pWin->optional &&
 	        CheckPassiveGrabsOnWindow(pWin, device, event, sendCore))
-		return TRUE;
+            {
+		ret = TRUE;
+                goto out;
+            }
 	}
 
 	if ((focus->win == NoneWin) ||
 	    (i >= device->spriteInfo->sprite->spriteTraceGood) ||
 	    ((i > checkFirst) &&
              (pWin != device->spriteInfo->sprite->spriteTrace[i-1])))
-	    return FALSE;
+	    goto out;
     }
 
     for (; i < device->spriteInfo->sprite->spriteTraceGood; i++)
@@ -3653,10 +3657,16 @@ CheckDeviceGrabs(DeviceIntPtr device, De
 	pWin = device->spriteInfo->sprite->spriteTrace[i];
 	if (pWin->optional &&
 	    CheckPassiveGrabsOnWindow(pWin, device, event, sendCore))
-	    return TRUE;
+	{
+	    ret = TRUE;
+            goto out;
+        }
     }
 
-    return FALSE;
+out:
+    if (ret == TRUE && event->type == ET_KeyPress)
+        device->deviceGrab.activatingKey = event->detail.key;
+    return ret;
 }
 
 /**
